### ISSUE TITLE
File IO support - customizable text encoding

### ISSUE THREAD

**Issue by WaynePhillipsEA** [WaynePhillipsEA/twinbasic#555]
This issue is a split from the main File IO issue (#279), where problems can be reported specifically for the new text Encoding feature of the file Open syntax.

### This feature is considered experimental at this stage.

The syntax is available in the latest releases of twinBASIC, from v0.13.27 onwards.    Please report any problems / feature requests relating to this feature in this issue.  Once this feature is not considered experimental, this issue will be closed.

`Open pathname For mode [Access access] [lock] [Encoding EncString] As [ # ] filenumber [ Len = reclength ]`

EncString can be a string literal / variable.  Allowed values are 

- "default"  (system ANSI codepage, default if not supplied)
- "utf-8"
- "utf-8 bom"
- "utf-16"
- "utf-16 bom"
- any windows codepage number, such as "1252", optionally prefixed with "windows", e.g. "windows-1252"
- plus all the usual suspects, as mentioned by @wqweto [https://github.com/WaynePhillipsEA/twinbasic/issues/279#issuecomment-980526433](url)

UTF16 is mostly supported, but **not** with the Input # statement (you'll get an error), but Line Input # works.  UTF-32 is not supported.

The "utf-8 bom" and "utf-16 bom" versions will output the appropriate BOM file marker when opening a file for output, and will verify the BOM marker when opening a file for input.

**Comment by rboeckka** [comment]
Could you make a small sample how it would work for conversion between UTF and ANSI? Which codebase is internaly used by TB?

**Comment by WaynePhillipsEA** [comment]
Here's a simple example:

```
        Open "Z:\unicodeTest.txt" For Output Encoding "utf-8 BOM" As #1
            Print #1, "Hello from twinBASIC! 😎"
        Close #1

        Dim myString As String
        Open "Z:\unicodeTest.txt" For Input Encoding "utf-8 BOM" As #1
            Line Input #1, myString
        Close #1
        
        Debug.Print "FILE DATA: " & myString
```

Without specifying utf-8 as the encoding, you'll get "?" in place of the Unicode emoji.

**Comment by bclothier** [comment]
Just a thought. Instead of using string literals, we should provide constants to use. e.g. `Encoding.encUtf8` a la Access' locale strings. 

Also because we now can handle BOM'd files it might be nice to have a function to report the possible encoding. With BOM, we can be assured that it's UTF-X BOM. Without BOM, it's anybody's guess as to what encoding is used but we can at least do few more things. For instance, XML usually come with `<?xml charset="UTF-8" ?>` and HTML also does something similar. I understand Microsoft uses (used?) an API to autodetect encoding based on the patterns used within the text file, so such function may be convenient for the unfortunate situations where we have a text file but we are not 100% sure what encoding it is using.

**Comment by WaynePhillipsEA** [comment]
> Just a thought. Instead of using string literals, we should provide constants to use. e.g. `Encoding.encUtf8` a la Access' locale strings.

Yup, that's a good idea.

> 
> Also because we now can handle BOM'd files it might be nice to have a function to report the possible encoding. With BOM, we can be assured that it's UTF-X BOM. Without BOM, it's anybody's guess as to what encoding is used but we can at least do few more things. For instance, XML usually come with `<?xml charset="UTF-8" ?>` and HTML also does something similar. I understand Microsoft uses (used?) an API to autodetect encoding based on the patterns used within the text file, so such function may be convenient for the unfortunate situations where we have a text file but we are not 100% sure what encoding it is using.

There is an API for that:  [https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa740986(v=vs.85)?redirectedfrom=MSDN](url).   I agree that it would be useful, but that API is not going to be portable, and so once we go cross-platform we'd have problems with that.

**Comment by bclothier** [comment]
Keep in mind that while we might deal with unicode internally (and we should be!) if the application is designed to load some files that the user provides to the application, we may not have much choice in the matter and assuming unicode might not be always possible, esp. if they use non-Unicode locale that could _look_ like Unicode and thus generate gibberish. I don't know enough whether Linux and Mac OS X still use non-unicode code pages. One'd hope that after having Unicode for last 20 years....

**Comment by WaynePhillipsEA** [comment]
With v0.13.32, constants for the common codepages have now been defined in the VBA.TextEncodingConstants module.  Intellisense support is included:

<img width="605" alt="tbEncodingIntellisense" src="https://user-images.githubusercontent.com/25738109/145840088-ef2b7c97-e8e4-427d-bd77-fe27c1b0ce96.png">

**Comment by Kr00l** [comment]
IMO the VBA.TextEncodingConstants do not feel VB-ish. I assume it's too late already but I can't offer something better I guess.
For example a text clipboard format is as `vbCFText`.
So that it mind it could be with prefix TE (for text encoding)
Example. `vbTE_utf_8`
Or: `enc_utf_8`

Or keep it like it is. Just wanted to mention.

**Comment by WaynePhillipsEA** [comment]
Not too late... I doubt a change on this would inconvenience many, yet.

**Comment by bclothier** [comment]
TBH, I've never really liked those prefixes. They add nothing and as the number of references increase, there will be more and more collisions over the prefixes. The real problem, IMO, is that they unnecessarily pollute the global namespace. I'd much rather use `TextEncoding.Utf8` than `vbUtf8`.

I usually refer to [Joel Spolsky's excellent article about why we should avoid System Hungarian notation](https://www.joelonsoftware.com/2005/05/11/making-wrong-code-look-wrong/). Though the prefix `vb` or such don't refer to types but rather to the library, I still consider it a form of System Hungarian notation because it doesn't add anything.

**Comment by WaynePhillipsEA** [comment]
We might not like them being prefixed, but as they form part of the standard library, we should try to make them fit in as best we can IMO.

**Comment by Kr00l** [comment]
Just a stupid question.
Can vs code (or tB) issue an intellisense of string literals after the Encoding keyword? (hard-coded list)
This would eliminate the need of enum constants.

**Comment by WaynePhillipsEA** [comment]
Yes, that's no problem

**Comment by Kr00l** [comment]
> Yes, that's no problem

Ah that's cool. That's whats actually only needed and I don't see where such an enum can be used elsewhere, so my vote would then go to dismiss VBA.TextEncodingConstants. :>

**Comment by bclothier** [comment]
Just to consider:

```
Public Sub MyFancyPantFileOpen( _
  FilePath As String, _
  TextEncoding As <What do I put here if I don't have an enum?> _
)  
...
  Open FilePath For Input Encoding TextEncoding #i
...
End Function
```

I'd need a way of knowing that I wasn't passed a nonsensical value into the parameter `TextEncoding`....

**Comment by Kr00l** [comment]
> Just to consider:
> 
> ```
> Public Sub MyFancyPantFileOpen( _
>   FilePath As String, _
>   TextEncoding As <What do I put here if I don't have an enum?> _
> )  
> ...
>   Open FilePath For Input Encoding TextEncoding #i
> ...
> End Sub
> ```
> 
> I'd need a way of knowing that I wasn't passed a nonsensical value into the parameter `TextEncoding`....

Solution would be that the Open..Encoding.. also accepts of course integer codepage values. Just the intellisense is stringified.
Your wrapper can then use a custom defined enum. And here comes a good side. That custom enum could only include those codepages which you need. E.g. 65001 for utf8 etc.

**Comment by wqweto** [comment]
> I'd need a way of knowing that I wasn't passed a nonsensical value into the parameter `TextEncoding`....

Btw, enums don't prevent that i.e. you can pass `-1` there and compiler will remain silent.

Originally the idea was to use a string for the encoding (specified as `[Encoding EncString]` in first post) the way it's implemented in FreeBasic.

I suppose they did like this becase Windows vs editions of Linux support different encodings, some of which are not known at compile-time and/or at all so cannot be included in an enum.

**Comment by Kr00l** [comment]
> > I'd need a way of knowing that I wasn't passed a nonsensical value into the parameter `TextEncoding`....
> 
> Btw, enums don't prevent that i.e. you can pass `-1` there and compiler will remain silent.
> 
> Originally the idea was to use a string for the encoding (specified as `[Encoding EncString]` in first post) the way it's implemented in FreeBasic.
> 
> I suppose they did like this becase Windows vs editions of Linux support different encodings, some of which are not known at compile-time and/or at all so cannot be included in an enum.

If that's true (i dont know) then its maybe better to allow strings only for cross-platform.

Of course in the wrapper you then need to pass a string (or string constant) instead of an enum.

**Comment by bclothier** [comment]
I agree that it should be a string. I guess the answer to the hypothetical scenario is a need for a function like `IsValidEncoding(Value As String) As Boolean` to make it easy to test and validate the encoding. Of course, handling the error is also a choice, too.
